### PR TITLE
[docs] Add doxygen comments for keypool classes

### PR DESCRIPTION
**Pull request by jnewbery**
Docs/move-only

Adds doxygen comments for the CKeyPool and CReserveKey objects. The way these work is pretty confusing and it's easy to overlook details (eg https://github.com/bitcoin/bitcoin/pull/15557#discussion_r271956393).

These are on the verbose side, but I think too much commenting is better than not enough. Happy to take feedback on what's an appropriate level.

### CODE REVIEW

**Review by promag — COMMENTED** [review]
:clap: very informative! Thanks! Concept ACK.

**Comment by laanwj** [comment]
Thanks for adding documentation!
ACK fb776446810e8f8d3cada1613a30c68d4d3a6fff

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
rebased

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -135,14 +135,61 @@ enum WalletFlags : uint64_t {
...
+ * Keypool keys are stored in the wallet/keystore's keymap. The keypool data is
+ * stored as a sets of indexes in the wallet (setInternalKeyPool,
+ * setExternalKeyPool and set_pre_split_keypool), and a map from the key to the
+ * index (m_pool_key_to_index). The CKeyPool object is used to
+ * serialize/deserialize the pool data to/from the database.
+ */
 class CKeyPool
 {
 public:
+    // The time at which the key was generated. Set in AddKeypoolPubKeyWithDB
```
Comment (jonatack) [inline src/wallet/wallet.h]: For my understanding: doc/developer-notes.md mentions using `//! Description before the member` syntax to describe member or variable use to be picked up by Doxygen. Is the `//` syntax used here with the intent to not be picked up, or for another reason?

**Review by jonatack — COMMENTED** [review]
Very helpful documentation! :100: Concept ACK

File: src/wallet/wallet.h (hunk 1 above)
Comment (jnewbery) [reply src/wallet/wallet.h]: No reason, I just got the doxygen syntax wrong. Thanks for pointing this out!

**Comment by jnewbery** [comment]
Fixed @jonatack's comment.

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -135,14 +135,61 @@ enum WalletFlags : uint64_t {
...
+ * With the introduction of HD wallets (commit: f1902510), the keypool
+ * essentially became an address look-ahead pool. Restoring old backups can no
+ * longer definitively lose funds as long as the addresses used were from the
+ * wallet's HD seed (since all private keys can be rederived from the seed).
+ * However, if many addresses were used since the backup, then the wallet may
+ * not know how far ahead in the HD chain to look for its addresses. The
+ * keypool is used to implement a 'gap limit'. The keypool maintains a set of
+ * keys (by default 1000) ahead of the last used key and scans for the
+ * addresses of those keys.  This avoids the risk of not seeing transactions
+ * involving the wallet's addresses, or re-using the same address.
```
Comment (jonatack) [inline src/wallet/wallet.h]: /s/, or/ or of/` (missing "of")

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -135,14 +135,61 @@ enum WalletFlags : uint64_t {
...
+ *
+ * HD-split added a second keypool (commit: 02592f4c). There is an external
```
Comment (jonatack) [inline src/wallet/wallet.h]: Beginner perspective: ran `git show 02592f4c` to see what "HD-split" meant more exactly. Perhaps "The wallet with HD chain split feature (HD-split) added..."

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -135,14 +135,61 @@ enum WalletFlags : uint64_t {
...
+ * keypool (for addresses to hand out) and an internal keypool (for change
+ * addresses).
+ *
+ * Keypool keys are stored in the wallet/keystore's keymap. The keypool data is
+ * stored as a sets of indexes in the wallet (setInternalKeyPool,
```
Comment (jonatack) [inline src/wallet/wallet.h]: Seems this ought to be either "a set of" or "sets of"

**Review by jonatack — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/15777/commits/292f0ee148bbe3b305aa8c1623355cee08a05867

Feel free to ignore the nits I mention. Very helpful documentation, particularly for someone new to the codebase and its history like myself. Kudos for the relevant commit hashes to dive in further.

**Comment by jnewbery** [comment]
Thanks for the review @jonatack . I've taken all your suggestions.

**Comment by jonatack** [comment]
Thanks, John. Re-ACK f1a77b0c5176306ca9f6f30211e32d3502ed4281, doc-only changes with respect to previous review.

**Comment by jb55** [comment]
ACK f1a77b0c5176306ca9f6f30211e32d3502ed4281

**Pull request merged**